plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/kvm/ha/KvmHaAgentClient.java

```diff
@@ -0,0 +1,346 @@
+    private static final String VM_COUNT = "count";
+    private static final String STATUS = "status";
+    private static final String CHECK_NEIGHBOUR = "check-neighbour";
+    private static final int WAIT_FOR_REQUEST_RETRY = 2;
+    private static final int MAX_REQUEST_RETRIES = 2;
+    private static final JsonParser JSON_PARSER = new JsonParser();
+    static final String HTTP_PROTOCOL = "http";
+    static final String HTTPS_PROTOCOL = "https";
+    private final static String APPLICATION_JSON = "application/json";
+    private final static String ACCEPT = "accept";
+
+    @Inject
+    private VMInstanceDao vmInstanceDao;
+
+    /**
+     *  Returns the number of VMs running on the KVM host according to Libvirt.
+     */
+    public int countRunningVmsOnAgent(Host host) {
+        String protocol = getProtocolString();
+        String url = String.format("%s://%s:%d", protocol, host.getPrivateIpAddress(), getKvmHaMicroservicePortValue(host));
+        HttpResponse response = executeHttpRequest(url);
+
+        if (response == null)
+            return ERROR_CODE;
+
+        JsonObject responseInJson = processHttpResponseIntoJson(response);
+        if (responseInJson == null) {
+            return ERROR_CODE;
+        }
+
+        return responseInJson.get(VM_COUNT).getAsInt();
+    }
+
+    /**
+     * Returns the HTTP protocol. It can be 'HTTP' or 'HTTPS' depending on configuration 'kvm.ha.webservice.ssl.enabled'
+     */
+    protected String getProtocolString() {
+        boolean KvmHaWebserviceSslEnabled = KVMHAConfig.KvmHaWebserviceSslEnabled.value();
+        String protocol = HTTP_PROTOCOL;
+        if (KvmHaWebserviceSslEnabled) {
+            protocol = HTTPS_PROTOCOL;
+        }
+        return protocol;
+    }
+
+    /**
+     * Returns the port from the KVM HA Helper according to the configuration 'kvm.ha.webservice.port'
+     */
+    protected int getKvmHaMicroservicePortValue(Host host) {
+        Integer haAgentPort = KVMHAConfig.KvmHaWebservicePort.value();
+        if (haAgentPort == null) {
+            LOGGER.warn(String.format("Using default kvm.ha.webservice.port: %s as it was set to NULL for the cluster [id: %d] from %s.",
+                    KVMHAConfig.KvmHaWebservicePort.defaultValue(), host.getClusterId(), host));
+            haAgentPort = Integer.parseInt(KVMHAConfig.KvmHaWebservicePort.defaultValue());
+        }
+        return haAgentPort;
+    }
+
+    /**
+     * Lists VMs on host according to vm_instance DB table. The states considered for such listing are: 'Running', 'Stopping', 'Migrating'.
+     * <br>
+     * <br>
+     * Note that VMs on state 'Starting' are not common to be at the host, therefore this method does not list them.
+     * However, there is still a probability of a VM in 'Starting' state be already listed on the KVM via '$virsh list',
+     * but that's not likely and thus it is not relevant for this very context.
+     */
+    public List<VMInstanceVO> listVmsOnHost(Host host) {
```

Comment:
Maybe this method could be outside the client with the DAO and keep `countRunningVmsOnAgent` on it?

Reply:
That makes sense, I will re-check this part of the implementation.
Thanks for the review, @nvazquez!